python/ExoticEngine/Solvers/ImpliedVol.py

```
@@ -0,0 +1,117 @@
+import abc
+from collections.abc import Callable
+from enum import Enum
+from inspect import isfunction
+from typing import final
+
+import numpy as np
+from scipy.stats import norm
+
+from ExoticEngine import MonteCarloPricer as Pricer
+
+
+class PUT_CALL(Enum):
+    PUT = "PUT"
+    CALL = "CALL"
+
+
+class InvertFunction:
+    """
+    Base class used for functions with 1 parameter
+    If f is multidimensional, then need to define derived class
+    """
+
+    @abc.abstractmethod
+    def __init__(self, function: Callable):
+        """Child classes must define a different constructor"""
+        assert isfunction(function)
+        self._func = function
+
+    @abc.abstractmethod
+    def f(self, x: float) -> float:
+        return self._func(x)
+
+    def derivative(self, x: float):
+        """not an abstract method - not all solvers require derivative"""
+        return None
```

Should this throw instead of return None if derived classes attempt to call it accidently?

---

Agree, will amend